docs/cn/2019-08-05-new-design-of-backup-restore.md

```diff
@@ -0,0 +1,233 @@
+
+### 异常处理
+
+异常处理备份期间的异常和 select * 一样，可分为两种可恢复和不可恢复，所有的异常都可以直接复用 TiDB 现有的机制。
+
+可恢复异常一般包含：
+
+- RegionError，一般由 region split/merge，not leader 造成。
+- KeyLocked，一般由事务冲突造成。
+- Server is busy，一般由于 TiKV 太忙造成。
+
+当发生这些异常时，备份的进度不会被打断。
+
+除了以上的其他错误都是不可恢复异常，发生后，它们会打断备份的进度。
+
+### 超出 GC 时间
+
+超出 GC 时间是说，需要备份的数据已经被 GC，这情况一般发生在增量备份上，会导致增量不完整。在发生这个错的时候，外部程序需要重新来一个全量备份。
+
+现在默认的 GC 时间是 10 分钟，实在是太短了，这意味着一旦开启全量备份，增量备份必须马上接上，而且运行的间隔必须在 10 分钟内。这个频率集群肯定受不了。这个问题不只备份会遇到， CDC 也会遇到。在开始备份或者 CDC 的集群上需要适当延长 GC 时间，比如延长到 6 小时或者 12 小时。
+
+### 恢复
+
+恢复所需的工作有以下几个：
+
+1. 创建需要恢复的 database 和 table
+2. 根据 table 和 SST 文件元信息，进行 Split & Scatter Region
+3. 将备份下来的 SST 文件按需读取到 TiKV 各节点
+4. 根据新 table 的 ID 对 SST 进行 Key Rewrite
+5. 将处理好的 SST 文件 Ingest 到 TiKV
+
+### Key Rewrite
+
+由于 TiDB 的数据在 TiKV 那保存的形式是：
+
+```
+Key: tablePrefix{tableID}_recordPrefixSep{rowID}
+Value: [col1, col2, col3, col4]
+
+Key: tablePrefix{tableID}_indexPrefixSep{indexID}_indexedColumnsValue
+Value: rowID
+```
+
+在 Key 中编码了 tableID，所以我们不能直接把备份下来的 SST 文件不经任何处理恢复到 TiKV 中，否则就有可能因为 tableID 对不上而导致数据错乱。
+
+为了解决该问题，我们必须在恢复前对 SST 文件进行 key 的改写，将原有的 tableID 替换为新创建的 tableID，同样的 indexID 也需要相同的处理。
+
+### Split & Scatter
+
+TiKV 对 Region 的大小是有限制的，默认为 96MB，超出该阈值则需要分裂（Split）。集群刚启动时，只有少量的 Region 的，我们在恢复的时候不能把所有数据都恢复到这些少量的 Region 中，所以需要提前将 Region 分裂并打散（Scatter）。
+
+由于备份 SST 文件是按照 Region 生成的，天然就有合适的大小和对应的数据范围，所以我们可以在根据各个 SST 文件中的范围对集群中的 Region 进行分裂。分裂完成后还需要打散新分裂出来的 Region，防止发生数据倾斜。
+
+### 恢复流程
+
+目前的 restore 可以先不考虑支持 sql，只提供一个恢复工具，且仅支持在一个没有数据的全新集群上恢复数据
+
+1. 使用备份的 schema 信息创建一个新表
+2. 拿到新表的 table id，使用 table id 重写数据的 key，上传重写后 key-value 到 tikv-importer 上
+3. tikv-importer 接收到数据之后，重新切分、打散数据
+4. tikv-importer 将文件传到对应 tikv-server 上，随后调用 ingest 导入 SST 文件
+
+这个方案的优点是实现简单，不需要修改任何的现有代码，缺点是速度较慢，当 importer 在恢复过程中挂掉并且无法再启动时要从头开始恢复。
+
+### 初步对接计划
+
+恢复对接大体两部分：
+1. github.com/tikv/impoter 
+   * 支持读取备份 SST
+   * key 重写，然后写入 importer 内部 engine
+      1. 初期是否需要支持 key 重写？能不能在全新集群上导入？
+   * 添加恢复 RPC，方便外部工具指导 importer 如何进行恢复，包括：
+      1. 恢复的 key value 范围
+      2. 备份数据的地址
+2. github.com/overvenus/br
+   * 添加恢复 RPC，指导 importer 恢复
+
+> 现在使用的方案已经不再使用 tikv-importer（它也预计会在 4.0 被移除），转而使用了 TiKV 一组能够下载/导入 SST 文件的新 API。相关的信息可以在[这里](./2019-11-05-design-of-reorganize-importSST-to-TiKV.md)和[这里](./2019-09-24-BR-and-lightning-reorganization.md)找到。
+>
+> 现在的步骤大概是这样：
+>
+> 1. 不变
+> 2. 通过备份的元数据构建 key-value 的范围，同时依照新表 ID 构建 rewirte rules。
+> 3. 依照 rewrite rules 和 key 范围分割并打散 regions。（这一步由 BR 而非 tikv 完成了）
+> 4. 调用 download 和 ingest API 完成导入。
+
+### 问题
+
+1. 不支持在线恢复
+
+  现阶段 importer 只能离线在全新集群上恢复，做不到在线恢复，这是不符合用户的预期的。这里的困境是 ingest SST 失败会导致各个副本之间的数据不一致。
+
+
+更新：已通过 [placement rule](https://pingcap.com/docs-cn/stable/how-to/configure/placement-rules/#placement-rules-使用文档) 实现资源隔离
+2. 不支持原集群恢复
+
+  原集群恢复涉及到各种 ID 的问题，最简单办法就是 rewrite key，把 key 中老 ID 替换成新 ID。这是 crdb 的方案，我们同样可以借鉴。
+
+更新：已通过 [Key rewrite](./2019-09-09-BR-key-rewrite-disscussion.md) 在理论上支持原集群恢复
```

Comment:
These updates need to be placed in block quotes.

Reply:
That is done now.